proposals/data-classes.md

```diff
@@ -0,0 +1,372 @@
+allow the data to shift contexts and allow other components to define their
+own behaviors.
+
+## What does it look like in C#?
+
+C# has a couple different ways to represent simple data, but I contend that
+all have fundamental problems.
+
+**Anonymous types**. Things look pretty good at first: you can name, it
+can read it, and you can compare it easily (equality is automatically
+defined). You can't modify it, though -- anonymous types are always immutable
+and you can't easily create a copy with only one change. The real problem
+is composibility. You can't use it as a real type anywhere outside the
+current method and you can't nest it in other data structures, except as
+an object.
+
+**Tuples**. Tuples are a lot like anonymous types. You can provide names,
+read the elements, modify them, and compare them. You can't make them
+immutable, but the real problem is in composition. Tuples aren't really
+an abstraction -- you describe the data structure in full in every
+place you use it. This makes it hard to expand tuples past a certain
+size and makes it difficult to compose with other data structures because
+you cannot refer to them by name.
+
+**Classes/Structs**. This is by far the most common representation of data
+in C#. A canonical example looks something like this:
+
+```C#
+public class LoginResource
+{
+    public string Username { get; set; }
+    public string Password { get; set; }
+    public bool RememberMe { get; set; }
+}
+```
+
+This feature provides names, for both the members and the data structure, it
+provides easy nominal composition, and is easily composible with all other
+data structures. It also provides a convenient syntax for creation by
+interacting directly with the named data, e.g.
+
+```C#
+var x = new LoginResource {
+    Username = "andy",
+    Password = password
+}
+```
+
+Unfortunately, there are still serious problems. There is no piecewise
+comparer implicitly defined for C# classes, so if you want simple data
+comparison, the real example looks like this:
+
+```C#
+using System;
+
+public class LoginResource : IEquatable<LoginResource>
+{
+    public string Username { get; set; }
+    public string Password { get; set; }
+    public bool RememberMe { get; set; } = false;
+
+    public override bool Equals(object obj)
+        => obj is LoginResource resource && Equals(resource);
+
+    public bool Equals(LoginResource other)
+    {
+        return other != null &&
+               Username == other.Username &&
+               Password == other.Password &&
+               RememberMe == other.RememberMe;
+    }
+
+    public override int GetHashCode()
+    {
+        var hashCode = -736459255;
+        hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(Username);
+        hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(Password);
+        hashCode = hashCode * -1521134295 + RememberMe.GetHashCode();
+        return hashCode;
```

Comment:
note: this is simpler now with HashCode.Combine.  (just to make the example code fair :)).